cmd/bpf2go/vfs.go

```diff
@@ -0,0 +1,224 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"slices"
+	"strings"
+)
+
+// vfs is LLVM virtual file system parsed from YAML file
+//
+// In a nutshell, it is a tree of "directory" nodes with leafs being
+// either "file" (a reference to file) or "directory-remap" (a reference
+// to directory).
+//
+// https://github.com/llvm/llvm-project/blob/llvmorg-18.1.0/llvm/include/llvm/Support/VirtualFileSystem.h#L637
+type vfs struct {
+	Version       int       `json:"version"`
+	CaseSensitive bool      `json:"case-sensitive"`
+	Roots         []vfsItem `json:"roots"`
+}
+
+type vfsItem struct {
+	Name             string      `json:"name"`
+	Type             vfsItemType `json:"type"`
+	Contents         []vfsItem   `json:"contents,omitempty"`
+	ExternalContents string      `json:"external-contents,omitempty"`
+
+	// specificity is used internally to resolve conflicts
+	//
+	// We populate vfs from go modules. Surprisingly, module file
+	// trees may overlap, e.g. "github.com/aws/aws-sdk-go-v2" and
+	// "github.com/aws/aws-sdk-go-v2/internal/endpoints/v2".
+	specificity int
+}
+
+type vfsItemType string
+
+const (
+	vfsFile           vfsItemType = "file"
+	vfsDirectory      vfsItemType = "directory"
+	vfsDirectoryRemap vfsItemType = "directory-remap"
+)
+
+// vfsAdd adds "directory-remap" entry for dir under vpath
+func vfsAdd(root *vfsItem, vpath, dir string) error {
+	return vfsAdd2(root, vpath, dir, vfsDirectoryRemap, len(vpath), defaultFS{})
+}
+
+type defaultFS struct{}
+
+func (defaultFS) Open(name string) (fs.File, error) {
+	return os.Open(name)
+}
+
+func (defaultFS) Stat(name string) (fs.FileInfo, error) {
+	return os.Stat(name)
+}
+
+func vfsAdd2(vi *vfsItem, vpath, path string, typ vfsItemType, specificity int, fs fs.FS) error {
+	for _, name := range strings.Split(vpath, "/") {
```

Comment:
It is Golang module URL being split, it only ever contains forward slashes. Maybe `path` instead of `filepath`? Unfortunately, I can't find anything helpful in there either. (Besides a function to extract a single component at a time, starting from the right side.)